File: countPoints.cpp
/*Your code solves:

* LeetCode 1828: Queries on Number of Points Inside a Circle

## Problem Statement

You are given:

* An array `points`, where `points[i] = [xi, yi]` represents the coordinates of the `i`th point on a 2D plane.
* An array `queries`, where `queries[j] = [xj, yj, rj]` represents a circle centered at `(xj, yj)` with radius `rj`.

For each query, determine how many points lie **inside or on the boundary** of the corresponding circle.

Return an array `answer`, where `answer[j]` is the number of points inside the `j`th circle.

---

### Example 1

**Input**

```text
points = [[1,3],[3,3],[5,3],[2,2]]

queries = [[2,3,1],[4,3,1],[1,1,2]]
```

**Output**

```text
[3,2,2]
```

**Explanation**

* Query `[2,3,1]`:

  * Circle centered at `(2,3)` with radius `1`.
  * Points inside: `(1,3)`, `(2,2)`, `(3,3)`.
  * Count = **3**.

* Query `[4,3,1]`:

  * Circle centered at `(4,3)` with radius `1`.
  * Points inside: `(3,3)`, `(5,3)`.
  * Count = **2**.

* Query `[1,1,2]`:

  * Circle centered at `(1,1)` with radius `2`.
  * Points inside: `(1,3)`, `(2,2)`.
  * Count = **2**.

---

### Example 2

**Input**

```text
points = [[1,1],[2,2],[3,3]]

queries = [[2,2,1]]
```

**Output**

```text
[3]
```

---

### Constraints

```text
1 <= points.length <= 500
points[i].length == 2
0 <= xi, yi <= 500

1 <= queries.length <= 500
queries[j].length == 3
0 <= xj, yj <= 500
1 <= rj <= 500
```

---

### Function Signature (C++)

```cpp
vector<int> countPoints(
    vector<vector<int>>& points,
    vector<vector<int>>& queries
);
```

Return a vector where each element represents the number of points inside or on the boundary of the corresponding query circle.
*/



#include <bits/stdc++.h>
using namespace std;

class  countPoints{
public:
    std::vector<int> Solution(std::vector<std::vector<int>>& points, std::vector<std::vector<int>>& queries) {

        std::vector<std::array<int, 2>> pts;
        for(int i = 0; i < points.size(); i++)
        {
            pts.push_back({points[i][0], points[i][1]});
        }
        std::sort(pts.begin(), pts.end());

        int n = queries.size();
        std::vector<int> res(n);
        for(int i = 0; i < n; i++)
        {
            int x = queries[i][0];
            int y = queries[i][1];
            int r = queries[i][2];
            auto it = std::lower_bound(pts.begin(), pts.end(), std::array<int, 2>({x - r, 0}));

            for(; it != pts.end() && (*it)[0] <= x + r; it++) {
                int dx = (*it)[0] - x;
                int dy = (*it)[1] - y;
                if(dx * dx + dy * dy <= r * r)
                {
                    res[i]++;
                }
            }
        }
        return res;
    }
};
    int main(){
    vector<vector<int>> quieries={{2,3,1},{4,3,1},{1,1,2}};
    vector<vector<int>> points={{1,3},{3,3},{5,3},{2,2}};
    countPoints c ;
    vector<int> noofp=c.Solution(points,quieries);
    for(int c:noofp)
    cout<< c<<" ";
    
    }

File: count_points.py

"""Your code solves:

* LeetCode 2103: Rings and Rods

## Problem Statement

There are **10 rods** labeled from `0` to `9`.

You are given a string `rings` of even length, where every **two characters** represent one ring:

* The **first character** is the color of the ring:

  * `'R'` = Red
  * `'G'` = Green
  * `'B'` = Blue
* The **second character** is the rod number (`'0'` to `'9'`) on which the ring is placed.

Return the number of rods that have **all three colors** of rings on them.

---

### Example 1

**Input**

```text
rings = "B0B6G0R6R0R6G9"
```

**Output**

```text
1
```

**Explanation**

The rings are:

| Ring | Rod |
| ---- | --- |
| B    | 0   |
| B    | 6   |
| G    | 0   |
| R    | 6   |
| R    | 0   |
| R    | 6   |
| G    | 9   |

* Rod `0` has **B, G, R** ✅
* Rod `6` has **B, R**
* Rod `9` has **G**

Only **one rod** contains all three colors.

---

### Example 2

**Input**

```text
rings = "B0R0G0R9R0B0G0"
```

**Output**

```text
1
```

---

### Example 3

**Input**

```text
rings = "G4"
```

**Output**

```text
0
```

---

### Constraints

```text
1 <= rings.length <= 100
rings.length is even.
rings consists only of the characters 'R', 'G', 'B', and digits '0' through '9'.
```

---

### Function Signature

```python
def countPoints(rings: str) -> int:
```

Return the number of rods that have **at least one Red, one Green, and one Blue ring**.


"""


from collections import defaultdict

class count_points:
    def solution(self, s: str) -> int:
        ans = 0
        r = defaultdict(str)
        print(r)
        for rod in range(1,len(s),2):
            ring = rod-1
            r[s[rod]] += s[ring]
            print(r)

        for key in r:
            if len(set(r[key])) >= 3:
                ans +=1

        return ans

if __name__ =="__main__":
     cp=count_points()
     rings="B0B6G0R6R0R6G9"
     print(cp.solution(rings))
